NorthstarDLL/mods/modsavefiles.cpp

```diff
@@ -0,0 +1,532 @@
+/// <returns>The size of the contents of the current directory, excluding a specific file.</returns>
+uintmax_t GetSizeOfFolderContentsMinusFile(fs::path dir, std::string file)
+{
+	uintmax_t result = 0;
+	for (const auto& entry : fs::directory_iterator(dir))
+	{
+		if (entry.path().filename() == file)
+			continue;
+		// fs::file_size may not work on directories - but does in some cases.
+		// per cppreference.com, it's "implementation-defined".
+		try
+		{
+			result += fs::file_size(entry.path());
+		}
+		catch (fs::filesystem_error& e)
+		{
+			if (entry.is_directory())
+			{
+				result += GetSizeOfFolderContentsMinusFile(entry.path(), "");
+			}
+		}
+	}
+	return result;
+}
+
+// Saves a file asynchronously.
+template <ScriptContext context> void SaveFileManager::SaveFileAsync(fs::path file, std::string contents)
+{
+	auto mutex = std::ref(mutexMap[file]);
+	std::thread writeThread(
+		[mutex, file, contents]()
+		{
+			try
+			{
+				mutex.get().lock();
+
+				fs::path dir = file.parent_path();
+				// this actually allows mods to go over the limit, but not by much
+				// the limit is to prevent mods from taking gigabytes of space,
+				// we don't need to be particularly strict.
+				if (GetSizeOfFolderContentsMinusFile(dir, file.filename().string()) + contents.length() > MAX_FOLDER_SIZE)
+				{
+					// tbh, you're either trying to fill the hard drive or use so much data, you SHOULD be congratulated.
+					spdlog::error(fmt::format("Mod spamming save requests? Folder limit bypassed despite previous checks. Not saving."));
+					mutex.get().unlock();
+					return;
+				}
+
+				std::ofstream fileStr(file);
+				if (fileStr.fail())
+				{
+					mutex.get().unlock();
+					return;
+				}
+
+				fileStr.write(contents.c_str(), contents.length());
+				fileStr.close();
+
+				mutex.get().unlock();
+				// side-note: this causes a leak?
+				// when a file is added to the map, it's never removed.
+				// no idea how to fix this - because we have no way to check if there are other threads waiting to use this file(?)
+				// tried to use m.try_lock(), but it's unreliable, it seems.
+			}
+			catch (std::exception ex)
+			{
+				spdlog::error("SAVE FAILED!");
+				mutex.get().unlock();
+				spdlog::error(ex.what());
+			}
+		});
+
+	writeThread.detach();
+}
+
+// Loads a file asynchronously.
+template <ScriptContext context> int SaveFileManager::LoadFileAsync(fs::path file)
+{
+	int handle = ++m_iLastRequestHandle;
+	auto mutex = std::ref(mutexMap[file]);
+	std::thread readThread(
+		[mutex, file, handle]()
+		{
+			try
+			{
+				mutex.get().lock();
+
+				std::ifstream fileStr(file);
+				if (fileStr.fail())
+				{
+					spdlog::error("A file was supposed to be loaded but we can't access it?!");
+
+					g_pSquirrel<context>->AsyncCall("NSHandleLoadResult", handle, false, "");
+					mutex.get().unlock();
+					return;
+				}
+
+				std::stringstream stringStream;
+				while (fileStr.peek() != EOF)
+					stringStream << (char)fileStr.get();
```

Comment:
This is unnecessarily inefficient; just do `stringStream << fileStr.rdbuf()`.

Reply:
fun fact, we do this in lots of places (namely, mod loading)

Reply:
ugh, feel free to submit a PR to change that, then

Reply:
oh, and if we do this for mod loading too, it'd probably be even better to do this:

```c++
std::ifstream st("whatever");
std::string str;
st.seekg(0, std::ios::end);   
str.reserve(st.tellg());
st.seekg(0, std::ios::beg);
str.assign((std::istreambuf_iterator<char>(st)), std::istreambuf_iterator<char>());
```

That'll prevent the string from being continuously re-allocated, which should be significantly faster and reduce the spikes in memory usage.

Could even put this in a helper function somewhere (and if we do that, we should also add proper error handling too).

Reply:
sometimes, I really do hate c++

Reply:
So, wait, do I change it or just leave it as-is and resolve this?

Reply:
> So, wait, do I change it or just leave it as-is and resolve this?

I'd say definitely at least do `stringStream << fileStr.rdbuf()` for now

idk about the other (longer) suggestion, not sure if @pg9182 was suggesting to do that for mod loading, here, or both?

Reply:
Yeah, it has the entire kitchen sink except for one of the knobs xd.

Reply:
Resolved in df0554a